### ISSUE TITLE
[clj] Implement str as an alias for ++

### ISSUE THREAD

**Issue by yurrriq** [lfe/lfe#269]
I think we could do this even more robustly and convert different types appropriately.

https://github.com/lfex/clj/issues/16

**Comment by skovsgaard** [comment]
Would it make sense to use `io_lib:printable_list` to distinguish from a regular cons? Or are you thinking of adding implicit conversion so for instance `(clj:str "a " 1 " b " 2)` would return `"a 1 b 2"`? If yes, is this gonna work for tuples and non-printable lists as well?

**Comment by rvirding** [comment]
What does the clojure function `str` do? Is there one? Is it reasonable for the `clj` module to behave in a similar fashion? There is a little support for binary string in LFE, the `#"..."` syntax so how would we handle that? How does clojure handle stringifying data objects? I am used to using `lfe_io:print1`.

**Comment by skovsgaard** [comment]
[This](https://github.com/clojure/clojure/blob/clojure-1.8.0/src/clj/clojure/core.clj#L533) explains it better than I can, I imagine. But basically, it's intended as a constructor for strings, either taking strings or values which can be converted to strings.

**Comment by skovsgaard** [comment]
I can't really work out if this solution is sufficient or it's just overly complicated, but before I make it a PR, there's a gist of my idea [here](https://gist.github.com/skovsgaard/c938c17279ad93505519f440b9d80019)

**Comment by skovsgaard** [comment]
@yurrriq I added another comment on the gist. I don't think we get notifications there.

**Comment by yurrriq** [comment]
Thanks, you're right, but I got this one :smile:

**Comment by skovsgaard** [comment]
Okay, great. I managed to find the docs on variadic macros in the meantime. Thanks for the explanation, though.

**Comment by yurrriq** [comment]
One additional point to mention about exported macros is that they need to "expand fully" and not rely on internal functions/macros in their expanded form. I'm not sure how to explain that better before coffee. Maybe @rvirding can chime in.

You can test by adding `(export-macro str)` to the `defmodule` form (or adding `str` to an existing `export-macro` form) and then calling `(clj:str ...)`.

**Comment by skovsgaard** [comment]
Thanks. I'll keep that in mind; it seems like something I might otherwise have stumbled at.

Does this mean that if I do like in the gist and define a `-stringify/1` helper for use as part of this, that won't work if I do it with a macro?

**Comment by yurrriq** [comment]
Right, if the macro expands to a form containing `-stringify` then it won't export properly.

**Comment by skovsgaard** [comment]
No problem. I think I might have to drop that one too. I've played around with it in the shell today and the `-stringify` I had behaves weirdly for anything that isn't a number. I might be able to bake the desired behavior into the macro though.

**Comment by skovsgaard** [comment]
I'm absolutely stumped as to why the following only works for integers:

``` lisp
(defmacro str args
  `(lists:flatmap
    (lambda (x)
      (case x
        (x (when (is_integer x))
           (integer_to_list x))
        (x (when (is_float x))
           (float_to_list x '(#(decimals 2))))
        (x (when (is_atom x))
           (atom_to_list x))
        (_ x)))
    ,args))
```

It works completely as intended for ints but everything else passed to it and it fails with `bad form application` which isn't telling me a whole lot.

Update: I'm apparently a dumbass who's just unquoting stuff at random. That last `,args` needs to be a `(list ,@args)`. That makes it work with the remaining types.

**Comment by yurrriq** [comment]
~~You can't do `(case x (x ...) ...)` in LFE so the first clause will always match.~~

**Edit**: See [Robert's comment](https://github.com/rvirding/lfe/issues/269#issuecomment-254070884).

**Comment by yurrriq** [comment]
Side note: you might want `match-lambda` here instead.

**Comment by skovsgaard** [comment]
Got it. I've got a working version now using `match-lambda`. There'll be a PR soon as I have figured out what my tests are complaining about.

**Comment by rvirding** [comment]
Not you have a guard so the whole pattern won't always match.

But I have worked out why. If you give the `args` as just integers `(str 1 2 3)` then `,args` expands to `(1 2 3)` which is a list integers which is interpreted as string and they are self quoting. It you include something else `(str 1 2.2 3)` then `,args` expands to `(1 2.2 3)` which is a normal function call and there is no function `1`.

When you change the macro to have `(list ,@args)` then it will expand to `(list 1 2.2 3)` which just builds the list, after expanding the elements.

**Comment by rvirding** [comment]
Using a `match-lambda` in this case just change that in this case but it is more typical LFE style of code.

**Comment by rvirding** [comment]
There is a functions `lists:concat/1` which will do that for integers, floats, atoms and strings, but the floats are full sized.

**Comment by skovsgaard** [comment]
I don't know why I settled on 2 digits to truncate to, but it's really the only way I could think of to get around the issue of very long floats when stringified.

**Comment by skovsgaard** [comment]
And thanks for the explanation. Even though I got around it, I was wondering what was going on where it only worked for ints.

**Comment by skovsgaard** [comment]
I opened a PR on the version using `match-lambda`. As for using `lists:concat/1`, would it make sense to add a note in the docstring about using that if passing a list of values is needed (as opposed to here where it's an arbitrary number of scalar values)?

**Comment by rvirding** [comment]
No, I would leave is as a macro which combines its arguments into a string. Let them find `lists:concat` by them selves as it is not part of the clj module.

**Comment by skovsgaard** [comment]
Would you prefer we had a macro like

``` lisp
(defmacro str args
  `(lists:concat (list ,@args)))
```

which would not be local to the `clj` module?

If yes, I really like making it that much simpler, but how would we deal with binaries if at all?

**Comment by yurrriq** [comment]
A fully-qualified call is fine, even it it's `clj:foo` (so long as `foo` is exported).

Can we use `lfe_io:print1` at all?

**Comment by yurrriq** [comment]
What about something like this?

``` lfe
(defmacro str args
  `(lists:flatmap #'lfe_io:print1/1 (list ,@args)))
```

**Comment by skovsgaard** [comment]
That last one I can't get to work. Does it need something else to make it printable? And what's the advantage of using `lfe_io:print1`?

**Comment by rvirding** [comment]
This was closed by #291
